### PR TITLE
Example of cumulative counting using updateStateByKey

### PR DESCRIPTION
**Pull request by skumargithub**
A simple example which uses updateStateByKey on a DStream
to perform a cumulative wordcount of words coming in from a network socket

### CODE REVIEW

File: examples/src/main/scala/spark/streaming/examples/NetworkWordCumulativeCountUpdateStateByKey.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * Usage: NetworkWordCumulativeCountUpdateStateByKey <master> <hostname> <port>
+ *   <master> is the Spark master URL. In local mode, <master> should be 'local[n]' with n > 1.
+ *   <hostname> and <port> describe the TCP server that Spark Streaming would connect to receive data.
+ *
+ * To run this on your local machine, you need to first run a Netcat server
+ *    `$ nc -lk 9999`
+ * and then run the example
+ *    `$ ./run spark.streaming.examples.NetworkWordCumulativeCountUpdateStateByKey local[2] localhost 9999`
+ */
+object NetworkWordCumulativeCountUpdateStateByKey {
```
Comment (tdas) [inline examples/src/main/scala/spark/streaming/examples/NetworkWordCumulativeCountUpdateStateByKey.scala]: Such long names are not encouraged. Can you please change it to something like StatefulNetworkWordCount (stateful put at the front to catch the eye of whoever want to implement something with state).

File: examples/src/main/scala/spark/streaming/examples/NetworkWordCumulativeCountUpdateStateByKey.scala [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  def main(args: Array[String]) {
+    if (args.length < 3) {
+      System.err.println("Usage: NetworkWordCountUpdateStateByKey <master> <hostname> <port>\n" +
+        "In local mode, <master> should be 'local[n]' with n > 1")
+      System.exit(1)
+    }
+
+    val updateFunc = (values: Seq[Int], state: Option[Int]) => {
+      val currentCount = values.foldLeft(0)(_ + _)
+      //println("currentCount: " + currentCount)
```
Comment (tdas) [inline examples/src/main/scala/spark/streaming/examples/NetworkWordCumulativeCountUpdateStateByKey.scala]: Please remove the commented lines. Its best to keep the example as concise as possible.

File: examples/src/main/scala/spark/streaming/examples/NetworkWordCumulativeCountUpdateStateByKey.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    val lines = ssc.socketTextStream(args(1), args(2).toInt)
+    val words = lines.flatMap(_.split(" "))
+    val wordDstream = words.map(x => (x, 1))
+
+    // Update the cumulative count using updateStateByKey
+    // This will give a Dstream made of state (which is the cumulative count of the words)
+    val stateDstream = wordDstream.updateStateByKey[Int](updateFunc)
+
+    stateDstream.foreach(rdd => {
+      rdd.foreach(rddVal => {
```
Comment (tdas) [inline examples/src/main/scala/spark/streaming/examples/NetworkWordCumulativeCountUpdateStateByKey.scala]: rdd.foreach will print the elements in the stdout of the worker nodes, which wont be visible if this example is run on a cluster. Rather, using stateDstream.print() (as shown in the NetworkWordCount example) is recommended.

**Comment by tdas** [comment]
Thank you very much for this much needed example of updateStateByKey. I have a few comments that should be easy to fix.

File: examples/src/main/scala/spark/streaming/examples/StatefulNetworkWordCount.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ *   <master> is the Spark master URL. In local mode, <master> should be 'local[n]' with n > 1.
+ *   <hostname> and <port> describe the TCP server that Spark Streaming would connect to receive data.
+ *
+ * To run this on your local machine, you need to first run a Netcat server
+ *    `$ nc -lk 9999`
+ * and then run the example
+ *    `$ ./run spark.streaming.examples.StatefulNetworkWordCount local[2] localhost 9999`
+ */
+object StatefulNetworkWordCount {
+  private def className[A](a: A)(implicit m: Manifest[A]) = m.toString
```
Comment (tdas) [inline examples/src/main/scala/spark/streaming/examples/StatefulNetworkWordCount.scala]: This className function is not being used anywhere as far as I see from the code. Is it really necessary?

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by pwendell** [comment]
This has been sitting around... I'm going to run tests again (could have been false alarm).

Jenkins, test this please

**Comment by mateiz** [comment]
Thanks for submitting this! I've merged it in.

**Comment by skumargithub** [comment]
Matei, Glad to be of some help!

**Pull request merged**